### PR TITLE
implement pause on system suspend

### PR DESCRIPTION
**Pull request by devsnek**
This is especially helpful on the steam deck, allowing you to safely press the power button like a real switch.

### CODE REVIEW

File: src/common/intrusive_red_black_tree.h [hunk 1]
```diff
@@ -247,14 +247,14 @@ concept HasRedBlackKeyType = requires {
...
-            return static_cast<Default*>(nullptr);
-        }
+template <typename T, typename Default>
+consteval auto* GetRedBlackKeyType() {
+    if constexpr (HasRedBlackKeyType<T>) {
+        return static_cast<typename T::RedBlackKeyType*>(nullptr);
+    } else {
+        return static_cast<Default*>(nullptr);
     }
+}
```
Comment (Morph1984) [inline src/common/intrusive_red_black_tree.h:257]: This unrelated change should probably be omitted

File: src/yuzu/main.cpp [hunk 2]
```diff
@@ -377,6 +377,24 @@ GMainWindow::GMainWindow()
...
 #endif
 
+#ifdef __linux__
+    {
+        auto bus = QDBusConnection::systemBus();
+        if (bus.isConnected()) {
+            bool success = bus.connect(
+                QStringLiteral("org.freedesktop.login1"), QStringLiteral("/org/freedesktop/login1"),
+                QStringLiteral("org.freedesktop.login1.Manager"), QStringLiteral("PrepareForSleep"),
+                QStringLiteral("b"), this, SLOT(OnPrepareForSleep(bool)));
```
Comment (Morph1984) [inline src/yuzu/main.cpp:387]:
```suggestion
                QStringLiteral("b"), this, &GMainWindow::OnPrepareForSleep);
```

File: src/common/intrusive_red_black_tree.h (hunk 1 above)
Comment (devsnek) [reply src/common/intrusive_red_black_tree.h:257]: its dirty from `ninja clang-format`. some pr has got to fix it eventually :P

File: src/yuzu/main.cpp [hunk 3]
```diff
@@ -377,6 +377,24 @@ GMainWindow::GMainWindow()
     SDL_EnableScreenSaver();
 #endif
 
+#ifdef __linux__
+    {
+        auto bus = QDBusConnection::systemBus();
+        if (bus.isConnected()) {
+            bool success = bus.connect(
```
Comment (Morph1984) [inline src/yuzu/main.cpp:384]:
```suggestion
            const bool success = bus.connect(
```

File: src/yuzu/main.cpp [hunk 4]
```diff
@@ -1285,6 +1303,23 @@ void GMainWindow::OnDisplayTitleBars(bool show) {
     }
 }
 
+void GMainWindow::OnPrepareForSleep(bool sleep) {
```
Comment (Morph1984) [inline src/yuzu/main.cpp:1306]: Perhaps use `prepare_sleep` instead? As sleep is a function defined in `unistd.h`

File: src/yuzu/main.cpp [hunk 5]
```diff
@@ -377,6 +377,24 @@ GMainWindow::GMainWindow()
     SDL_EnableScreenSaver();
 #endif
 
+#ifdef __linux__
```
Comment (Morph1984) [inline src/yuzu/main.cpp:380]: Also this entire block should likely be extracted into its own function, and moved below the Common::Log::Start() call

File: src/yuzu/main.cpp (hunk 2 above)
Comment (devsnek) [reply src/yuzu/main.cpp:387]: this does not compile

Comment (Morph1984) [reply src/yuzu/main.cpp:387]: Ah, interestingly enough QDBusConnection still uses the old slot syntax, disregard this review then

File: src/yuzu/main.h [hunk 6]
```diff
@@ -252,6 +255,7 @@ private slots:
     void OnPauseGame();
     void OnPauseContinueGame();
     void OnStopGame();
+    void OnPrepareForSleep(bool sleep);
```
Comment (Kewlan) [inline src/yuzu/main.h:258]:
```suggestion
    void OnPrepareForSleep(bool prepare_sleep);
```
Missed renaming this parameter?

**Review by liamwhite — APPROVED** [review]
